File: scripts/convert-to-webp.py
import os
from PIL import Image

public_dir = os.path.join(os.path.dirname(__file__), '..', 'public')
exts = {'.jpg', '.jpeg', '.png', '.gif', '.bmp', '.tiff', '.avif'}

def collect_images(directory):
    for root, _, files in os.walk(directory):
        for f in files:
            if os.path.splitext(f)[1].lower() in exts:
                yield os.path.join(root, f)

for input_path in collect_images(public_dir):
    output_path = os.path.splitext(input_path)[0] + '.webp'
    if os.path.exists(output_path):
        print(f'Skip (already exists): {os.path.relpath(output_path, public_dir)}')
        continue
    img = Image.open(input_path).convert('RGB')
    img.save(output_path, 'WEBP', quality=85)
    before = os.path.getsize(input_path) // 1024
    after = os.path.getsize(output_path) // 1024
    print(f'Converted: {os.path.relpath(input_path, public_dir)} -> .webp  ({before}KB -> {after}KB)')

File: scripts/convert-to-webp.js
const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const publicDir = path.resolve(__dirname, '..', 'public');

function collectImages(dir, exts) {
  const results = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...collectImages(full, exts));
    } else if (exts.includes(path.extname(entry.name).toLowerCase())) {
      results.push(full);
    }
  }
  return results;
}

async function convert() {
  const images = collectImages(publicDir, ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.tiff', '.avif']);

  for (const input of images) {
    const output = input.replace(/\.(jpg|jpeg|png|gif|bmp|tiff|avif)$/i, '.webp');
    if (fs.existsSync(output)) {
      console.log(`Skip (already exists): ${path.relative(publicDir, output)}`);
      continue;
    }
    await sharp(input).webp({ quality: 85 }).toFile(output);
    const before = (fs.statSync(input).size / 1024).toFixed(0);
    const after = (fs.statSync(output).size / 1024).toFixed(0);
    console.log(`Converted: ${path.relative(publicDir, input)} → .webp  (${before}KB → ${after}KB)`);
  }
}

convert().catch(console.error);
